wallet/src/main/java/ch/admin/bag/covidcertificate/wallet/transfercode/logic/TransferCodeCrypto.kt

```diff
@@ -17,21 +17,38 @@ import android.security.keystore.KeyInfo
 import android.security.keystore.KeyProperties
 import ch.admin.bag.covidcertificate.sdk.core.extensions.fromBase64
 import ch.admin.bag.covidcertificate.sdk.core.extensions.toBase64
+import ch.admin.bag.covidcertificate.wallet.data.WalletSecureStorage
+import org.bouncycastle.jce.provider.BouncyCastleProvider
 import java.security.KeyFactory
 import java.security.KeyPair
 import java.security.KeyPairGenerator
 import java.security.KeyStore
+import java.security.Security
 import java.security.Signature
+import java.security.spec.MGF1ParameterSpec
+import java.security.spec.PKCS8EncodedKeySpec
 import java.security.spec.RSAKeyGenParameterSpec
+import java.security.spec.X509EncodedKeySpec
 import java.time.Instant
 import javax.crypto.Cipher
 import javax.crypto.spec.GCMParameterSpec
+import javax.crypto.spec.OAEPParameterSpec
+import javax.crypto.spec.PSource.PSpecified
 import javax.crypto.spec.SecretKeySpec
 
 
 object TransferCodeCrypto {
 
 	const val ANDROID_KEYSTORE_NAME = "AndroidKeyStore"
+	const val BOUNCY_CASTLE_PROVIDER = "BC"
+
+	init {
+		if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N) {
+			// Remove platform provided bouncy castle provider and add updated one from library on Android 23 and lower
+			Security.removeProvider(BOUNCY_CASTLE_PROVIDER)
+			Security.addProvider(BouncyCastleProvider())
+		}
+	}
```

Comment:
Do we want to do this "globally" here, or explicitly for all functions? (e.g. `KeyPairGenerator.getInstance(KeyProperties.KEY_ALGORITHM_RSA, BOUNCY_CASTLE_PROVIDER)`)

Reply:
If you don't add the provider, you can't get it even if you specify it explicitly in the `getInstance()` methods. The `provider` parameter in those methods only specifies the provider you're requesting, it still has to be available in the `Security.getProviders()` list, which by default only contains the outdated BouncyCastle provider from the Android platform